### PR TITLE
STRF-4237: Split off rendering functionality

### PR DESCRIPTION
**Pull request by mattolson**
* Split off actual rendering functionality into new module called [paper-handlebars](https://github.com/bigcommerce/paper-handlebars), which encapsulates the use of Handlebars for rendering. This split allows us to introduce Handlebars v4 and potentially other template engines in the future, without changing the Paper interface.
* Remove helpers (these have been moved to paper-handlebars).
* Add `rendererType` parameter to Paper constructor. This defaults to `handlebars` (the current v3 implementation). If you pass `handlebars-v4`, it will use 4.0.11.
* Add Node 8.x to Travis
* Updated package.json to use `~` instead of `^` and updated dependencies to match what we're currently using.

## Breaking changes
* Removed the direct access of `contentServiceContext` for setting page content. From now on, use `addContent()` and `getContent()`.
* Removed `getTemplateProcessor()`. This is an internal concern of `paper-handlebars` and is used by `loadTemplates`.
* Removed `loadTemplatesSync()`. This was only used by helper tests and is no longer needed.
* Removed `handlebars` instance variable. Hopefully nobody is accessing that directly. Any helpers that were accessing it have been updated in `paper-handlebars` to use the global context they are given rather than accessing Paper directly at all.
* The `translator` attribute has been moved to `paper-handlebars` and is no longer accessible directly on Paper.
* The `decorators` attribute has been moved to `paper-handlebars` and is no longer accessible directly on Paper.
* The `settings` attribute has been renamed to `siteSettings`. This should only be accessed by `paper-handlebars`.
* The `cdnify()` function has been moved into a helper library in `paper-handlebars`.
* The `inject` attribute has been removed. This is storage used by two of the helpers, and the implementation has moved to `paper-handlebars`.

### CODE REVIEW

**Comment by mattolson** [comment]
@bigcommerce/storefront-team

File: index.js [hunk 1]
```diff
@@ -59,234 +48,161 @@ Fs.readdirSync(Path.join(__dirname, 'helpers')).forEach(function (file) {
...
         this.assembler = assembler || {};
-        this.contentServiceContext = {};
 
-        helpers.forEach(helper => helper(this));
+        // Build renderer based on type
+        switch(rendererType) {
+        case 'handlebars-v4':
+            this.renderer = new HandlebarsRenderer(this.siteSettings, this.themeSettings, 'v4');
+            break;
+        case 'handlebars':
```
Comment (junedkazi) [inline index.js]: Can we just use `default` 🍹

File: index.js [hunk 2]
```diff
@@ -59,234 +48,161 @@ Fs.readdirSync(Path.join(__dirname, 'helpers')).forEach(function (file) {
...
-     * @param  {Function} callback
+     * Use the assembler to fetch partials/templates, then load them
+     * into the renderer.
+     *
+     * @param {String} path The root template path to load. All dependencies will be loaded as well.
+     * @param {Function} callback
     */
     loadTemplates(path, callback) {
-        let processor = this.getTemplateProcessor();
+        let preProcessor = this.renderer.getPreProcessor();
```
Comment (junedkazi) [inline index.js]: Can we move this call to the constructor as a variable so we don't make this call for each path in the list ?

File: index.js (hunk 1 above)
Comment (mattolson) [reply index.js]: I'd prefer to keep this explicit, even just as documentation

**Comment by mattolson** [comment]
@bigcommerce/storefront-team ♻️

File: package.json [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-  "version": "2.0.7",
-  "description": "A stencil plugin to register partials and helpers from handlebars and returns the compiled version for the stencil platform.",
+  "version": "3.0.0-rc.1",
+  "description": "A Stencil plugin to load template files and render pages using backend renderer plugins.",
   "main": "index.js",
   "author": "Bigcommerce",
   "license": "BSD-4-Clause",
   "scripts": {
-    "test": "lab -l -t 90"
+    "linter": "eslint .",
```
Comment (mjschock) [inline package.json]: yay!  i love the minimalism of using npm scripts over grunt/gulp tasks.  :)

File: index.js [hunk 4]
```diff
@@ -59,237 +48,172 @@ Fs.readdirSync(Path.join(__dirname, 'helpers')).forEach(function (file) {
...
+    constructor(siteSettings, themeSettings, assembler, rendererType) {
+        this.siteSettings = siteSettings || {};
         this.themeSettings = themeSettings || {};
         this.assembler = assembler || {};
-        this.contentServiceContext = {};
 
-        helpers.forEach(helper => helper(this));
-    }
+        // Build renderer based on type
+        switch(rendererType) {
```
Comment (mjschock) [inline index.js]: 🍹 part of me wants to reduce this to something like:

```
this.renderer = new HandlebarsRenderer(this.siteSettings, this.themeSettings, rendererType.split(/-/).pop();
```

but being explicit is probably a better idea

File: .travis.yml [hunk 5]
```diff
@@ -3,9 +3,7 @@ sudo: false
 node_js:
   - 4
   - 6
-  
+  - 8
```
[the remark this answers is not in the document]
Comment (mattolson) [reply .travis.yml]: We still need to support CLI, which supports Node 4 still.

**Pull request merged**